src/Microsoft.DotNet.Tools.Builder/CompileContext.cs

```diff
@@ -0,0 +1,248 @@
+            var contextsToCheck = new List<ProjectContext> {_rootProject};
+
+            //convert ProjectDescription to ProjectContext
+            var dependencyContexts = _dependencies.Select
+                (keyValuePair => ProjectContext.Create(keyValuePair.Value.Project.ProjectFilePath, _rootProject.TargetFramework));
+
+            contextsToCheck.AddRange(dependencyContexts);
+
+
+            return contextsToCheck;
+        }
+
+        private void CheckPATHProbing(ProjectContext project, IncrementalPreconditions preconditions)
+        {
+            var pathCommands = CompilerUtil.GetCommandsInvokedByCompile(project)
+                .Select(commandName => Command.Create(commandName, "", project.TargetFramework))
+                .Where(c => Command.CommandResolutionStrategy.Path.Equals(c.CommandResolution));
+
+            foreach (var pathCommand in pathCommands)
+            {
+                preconditions.AddPathProbingPrecondition(project.ProjectName(), pathCommand.CommandName);
+            }
+        }
+
+        private void CollectCompilerNamePreconditions(ProjectContext project, IncrementalPreconditions preconditions)
+        {
+            var projectCompiler = CompilerUtil.ResolveCompilerName(project);
+
+            if (!Constants.KnownCompilers.Any(knownCompiler => knownCompiler.Equals(projectCompiler, StringComparison.Ordinal)))
+            {
+                preconditions.AddUnknownCompilerPrecondition(project.ProjectName(), projectCompiler);
+            }
+        }
+
+        private void CollectScriptPreconditions(ProjectContext project, IncrementalPreconditions preconditions)
+        {
+            var preCompileScripts = project.ProjectFile.Scripts.GetOrEmpty(ScriptNames.PreCompile);
+            var postCompileScripts = project.ProjectFile.Scripts.GetOrEmpty(ScriptNames.PostCompile);
+
+            if (preCompileScripts.Any())
+            {
+                preconditions.AddPrePostScriptPrecondition(project.ProjectName(), ScriptNames.PreCompile);
+            }
+
+            if (postCompileScripts.Any())
+            {
+                preconditions.AddPrePostScriptPrecondition(project.ProjectName(), ScriptNames.PostCompile);
+            }
+        }
+
+        private bool InvokeCompileOnDependency(ProjectDescription projectDependency, string outputPath, string intermediateOutputPath)
+        {
+            var compileResult = Command.Create("dotnet-compile",
+                $"--framework {projectDependency.Framework} " +
+                $"--configuration {_args.ConfigValue} " +
+                $"--output \"{outputPath}\" " +
+                $"--temp-output \"{intermediateOutputPath}\" " +
+                (_args.NoHostValue ? "--no-host " : string.Empty) +
+                $"\"{projectDependency.Project.ProjectDirectory}\"")
+                .ForwardStdOut()
+                .ForwardStdErr()
+                .Execute();
+
+            return compileResult.ExitCode == 0;
+        }
```

Comment:
Why invoke dotnet-compile directly? Why not as dotnet compile? When native support comes, we can add a flag --native.